### PR TITLE
Begin replacing propertyInfo.* access with functions

### PR DESCRIPTION
**Pull request by gaearon**
The goal is to remove the "injection" step at initialization. I find it confusing because we always have to look in three places: the configs, the initialization logic, and then the use of those properties. I think this is partly why the logic got so tangled by now.

I'm starting by removing the `propertyInfo.*` properties one by one, replacing them with function calls. There might be better implementations—happy to revise them. In some cases it may seem like we'd have to repeat an attribute name a few more times than before, but I think it probably gzips well, and in any case I'll make sure the size is smaller or the same before merging.

Note I'm not changing logic yet. There are a few oddities in how code paths differ based on whether `propertyInfo` exists—which ideally shouldn't be the case. I'll looking at that later.

### CODE REVIEW

File: packages/react-dom/src/shared/DOMProperty.js [hunk 1]
```diff
@@ -255,25 +278,16 @@ var HTMLDOMPropertyConfig = {
     // See http://schema.org/docs/gs.html
     itemScope: HAS_BOOLEAN_VALUE,
     // These attributes must stay in the white-list because they have
-    // different attribute names (see DOMAttributeNames below)
+    // different attribute names (see `attributeNames`)
+    // TODO: this doesn't seem great? Probably means we depend on
+    // existence in the whitelist somewhere we shouldn't need to.
```
Comment (nhunzaker) [inline packages/react-dom/src/shared/DOMProperty.js]: I remember when we hit this adding custom attributes, and it didn't feel great. It would be awesome to get to the bottom of this.

Comment (gaearon) [reply packages/react-dom/src/shared/DOMProperty.js]: Yea, definitely will. I have another set of commits that dives into this but it turned out to be a rabbit hole and I decided to do these first.

File: packages/react-dom/src/shared/DOMProperty.js [hunk 2]
```diff
@@ -228,8 +251,8 @@ var HTMLDOMPropertyConfig = {
     loop: HAS_BOOLEAN_VALUE,
     // Caution; `option.selected` is not updated if `select.multiple` is
     // disabled with `removeAttribute`.
-    multiple: MUST_USE_PROPERTY | HAS_BOOLEAN_VALUE,
-    muted: MUST_USE_PROPERTY | HAS_BOOLEAN_VALUE,
+    multiple: HAS_BOOLEAN_VALUE,
+    muted: HAS_BOOLEAN_VALUE,
```
Comment (nhunzaker) [inline packages/react-dom/src/shared/DOMProperty.js]: I think this is probably fine, but why the change?

Comment (gaearon) [reply packages/react-dom/src/shared/DOMProperty.js]: Not sure what you mean? I removed the flag. https://github.com/facebook/react/pull/11733/commits/c9ca797dfa3e3522fd3fcb432e3ffb70b15d0215

Easier to read as individual commits.

File: packages/react-dom/src/client/DOMPropertyOperations.js [hunk 3]
```diff
@@ -159,13 +162,13 @@ export function setValueForProperty(node, name, value) {
...
       // `toString`ed by IE8/9.
-      node[propertyInfo.propertyName] = value;
+      node[name] = value;
     } else {
-      var attributeName = propertyInfo.attributeName;
-      var namespace = propertyInfo.attributeNamespace;
+      var attributeName = getAttributeName(name);
+      var namespace = getAttributeNamespace(name);
       // `setAttribute` with objects becomes only `[object]` in IE8/9,
       // ('' + value) makes it output the correct toString()-value.
```
Comment (nhunzaker) [inline packages/react-dom/src/client/DOMPropertyOperations.js]: Is this true?

screen shot 2017-11-30 at 8 02 51 pm

Comment (nhunzaker) [reply packages/react-dom/src/client/DOMPropertyOperations.js]: Also valueOf:

screen shot 2017-11-30 at 8 03 38 pm

Comment (gaearon) [reply packages/react-dom/src/client/DOMPropertyOperations.js]: Maybe only IE8?

Comment (gaearon) [reply packages/react-dom/src/client/DOMPropertyOperations.js]: (Regardless, don't want to change the logic here, only the structure :-)

Comment (nhunzaker) [reply packages/react-dom/src/client/DOMPropertyOperations.js]: Totally. It's just nice to look at this code again with a fresh set of eyes. I'll file an issue.

File: packages/react-dom/src/shared/DOMProperty.js [hunk 4]
```diff
@@ -201,6 +176,54 @@ export function isReservedProp(name) {
...
+    default:
+      return null;
+  }
+}
+
+const usePropertiesFor = new Set(['checked', 'multiple', 'muted', 'selected']);
+
+export function shouldUseProperty(name) {
+  return usePropertiesFor.has(name);
+}
```
Comment (aweary) [inline packages/react-dom/src/shared/DOMProperty.js]: Could this just be inlined in `DOMPropertyOperations`? Doesn't look like it's used anywhere else. Also, we could then just do check the set directly instead of a function call too.

File: packages/react-dom/src/shared/DOMProperty.js (hunk 2 above)
Comment (nhunzaker) [reply packages/react-dom/src/shared/DOMProperty.js]: Got it. Thank you!

File: packages/react-dom/src/shared/DOMProperty.js (hunk 4 above)
Comment (gaearon) [reply packages/react-dom/src/shared/DOMProperty.js]: I want to consolidate all property/attribute stuff here first, then refactor it to make function contracts clearer, then possibly spread it out. I'd rather not spread it out yet.

File: packages/react-dom/src/client/DOMPropertyOperations.js (hunk 3 above)
Comment (nhunzaker) [reply packages/react-dom/src/client/DOMPropertyOperations.js]: https://github.com/facebook/react/issues/11735

File: packages/react-dom/src/shared/DOMProperty.js (hunk 4 above)
Comment (gaearon) [reply packages/react-dom/src/shared/DOMProperty.js]: It's just easier for me to deal with it if it's all in one file. Easier to see patterns.

File: packages/react-dom/src/shared/DOMProperty.js (hunk 2 above)
Comment (aweary) [reply packages/react-dom/src/shared/DOMProperty.js]: @gaearon at this point it looks like there aren't any properties that contain more than one flag. We could probably stop using bitmasks and just use an enum, remove `checkMask` and use a single property in `propertyInfo` instead of 6.

Comment (gaearon) [reply packages/react-dom/src/shared/DOMProperty.js]: My plan is to get rid of `propertyInfo` completely (maybe). This PR is just very iterative.

Comment (gaearon) [reply packages/react-dom/src/shared/DOMProperty.js]: This is a good observation though.

**Comment by gaearon** [comment]
@aweary Indeed it's much nicer without masks, thanks for the tip.

File: packages/react-dom/src/shared/DOMProperty.js [hunk 5]
```diff
@@ -20,82 +20,42 @@ var RESERVED_PROPS = {
...
+// An attribute that must be numeric or parse as a numeric and should be
+// removed when set to a falsey value.
+const NUMERIC = 3;
+// An attribute that must be positive numeric or parse as a positive
+// numeric and should be removed when set to a falsey value.
+const POSITIVE_NUMERIC = 4;
+// An attribute that can be used as a flag as well as with a value.
+// Removed when strictly equal to false; present without a value when
+// strictly equal to true; present with a value otherwise.
+const OVERLOADED_BOOLEAN = 5;
```
Comment (nhunzaker) [inline packages/react-dom/src/shared/DOMProperty.js]: :+1: :100:

**Comment by aweary** [comment]
@gaearon glad it worked out, it looks great!

At this point, I wonder if we could just get rid of `HTMLDOMPropertyConfig` and initialize the `propertyTypes` map with those values? Then we can get rid of `injectDOMPropertyConfig` in favor of something SVG-specific since those are the only values that are dynamically generated at runtime.

We'd lose out on the `injectDOMPropertyConfig` warning, but I'm not sure if it has a lot of value anyways since we're not exporting injection logic right?

**Comment by gaearon** [comment]
>At this point, I wonder if we could just get rid of HTMLDOMPropertyConfig and initialize the propertyTypes map with those values?

Yes but we'll need to see what's more compact in terms of bundle size. I'm not sure.

**Comment by aweary** [comment]
> Yes but we'll need to see what's more compact in terms of bundle size. I'm not sure.

@gaearon I checked this out locally  [and made the change](https://github.com/aweary/react/commit/e6e8ea3821a0853e259654254b44bf546701d73f) to compare bundle sizes. 

By inlining the config the `NODE_PROD` build for `react-dom.production.min.js` went from 29.39kb to 29.34kb, and the `UMD` build went from 30.49kb to 30.46kb. So inlining will also reduce the bundle size a little!

**Comment by gaearon** [comment]
Nice. Please push to my branch?

File: packages/react-dom/src/shared/DOMProperty.js [hunk 6]
```diff
@@ -168,23 +149,27 @@ export function shouldSetAttribute(name, value) {
...
   }
-  let propertyInfo = getPropertyInfo(name);
-  if (propertyInfo) {
-    return (
-      propertyInfo.hasBooleanValue ||
-      propertyInfo.hasStringBooleanValue ||
-      propertyInfo.hasOverloadedBooleanValue
-    );
+  if (propertyTypes.has(name)) {
+    switch (propertyTypes.get(name)) {
```
Comment (nhunzaker) [inline packages/react-dom/src/shared/DOMProperty.js]: Is the overhead of the switch greater than the double lookup of `has` and `get`?

File: packages/react-dom/src/shared/DOMProperty.js [hunk 7]
```diff
@@ -197,85 +182,98 @@ export function shouldAttributeAcceptBooleanValue(name) {
...
+      return false;
+  }
+}
+
+const usePropertiesFor: Set<string> = new Set([
+  'checked',
+  'multiple',
+  'muted',
+  'selected',
+]);
```
Comment (nhunzaker) [inline packages/react-dom/src/shared/DOMProperty.js]: I know microbenchmarks are dangerous, but would indexOf or object property membership be faster?

https://esbench.com/bench/5a21e22199634800a03494ec

screen shot 2017-12-01 at 6 14 23 pm

File: packages/react-dom/src/shared/DOMProperty.js (hunk 6 above)
Comment (gaearon) [reply packages/react-dom/src/shared/DOMProperty.js]: I was trying to avoid making it polymorphic (number/undefined). Maybe that's silly, dunno.

File: packages/react-dom/src/shared/DOMProperty.js (hunk 7 above)
Comment (aweary) [reply packages/react-dom/src/shared/DOMProperty.js]: I bet in the long-term checking set membership will be faster, since that's the sole purpose of the data structure. If we're still getting 25 million ops/s with a set it's probably fine to use it since its the correct data structure for this.

Comment (nhunzaker) [reply packages/react-dom/src/shared/DOMProperty.js]: Ya. I hesitated to say anything. Dan just mentioned the extra look-ups were slowing things down a bit.

Feel free to disregard.

Comment (aweary) [reply packages/react-dom/src/shared/DOMProperty.js]: If it's a non-trivial performance regression we should definitely consider changing it. It's just so nice using `Set` and `Map` instead of arrays and objects 😍

**Comment by gaearon** [comment]
Meh. It's too slow. I need to rethink this.

**Comment by gaearon** [comment]
https://github.com/facebook/react/pull/11815
